### ISSUE TITLE
Feature Request: add option for collapsing cmdline when empty.

### ISSUE THREAD

**Issue by parkerault** [vim/vim#940]
I would like to be able to hide the cmdline when not in use:

![screen shot 2016-07-21 at 1 28 38 pm-1](https://cloud.githubusercontent.com/assets/78856/17067266/30558e24-5008-11e6-94ab-cf8c51a824b4.jpg)

**Comment by chdiza** [comment]
But then look what would happen if some plugin were to do `:echo "foo"`.  The statusline would have to then jump up  by one line, only to go back down when something erases the echoed text.  This would happen every time anything echoed a one-line string, which happens a lot, especially for people with plugins installed.  You'd get a very jumpy statusline.

The same would apply to the message that vim prints in that spot when you simply save a file, e.g., `"myfile" 180L, 6025C written`.

In short, the space occupied by the cmdline is also used for other important things when the cmdline is not actively being typed in, most especially messages and echos.  Having the statusline jump up and down making room for these other things would be pretty annoying.

**Comment by parkerault** [comment]
That's probably true, but for those of us that wouldn't mind the jumpiness, `set cmdheight=0` would be a simple way to configure it.

**Comment by decaz** [comment]
It is very nice proposal to show cmdline only when you need it.
@chdiza, if you set cmdline=0 then you are sure you don't need to see echoes and want to open cmdline only when you press : or / so there is no any annoying jumping.
Same issue for Neovim: https://github.com/neovim/neovim/issues/1004

**Comment by purpleP** [comment]
To avoid jumpiness you can do one of two things. Show either statusline or commandline at the same time at the same line (I think that's how tmux handles it). Or show command line above the statusline. I large number of vim users uses no plugins or a not so many plugins (like me). My command line is almost always empty and just eats up space from the text. What got me in vim was it's minimalism. I could concentrate on the code without IDE's distractions. Command line that is showed always is making this ideal minimalistic editing just one step away. So close...

**Comment by chrisbra** [comment]
if you disable the statusline (and only have one window open) you basically gain one additional line. Other than that, I am tempted to **won't fix** this issue, there are a lot more important things open and I don't see the use case of this. It will just make it more confusing for users.

**Comment by purpleP** [comment]
@chrisbra I'm not pressing on this, and there are more important issues, but I think you're wrong about two things

1. "it will just make it more confusing". For new users help system is already confusing. That's why there is `help help` and stuff. For experienced user one more options (out of hundreds of them) wouldn't make any difference. And it may just be the same option just with allowing new value for it `cmdheight=0`.
2. You say you don't see the use case for that, but me and other people have described it very clearly. The fact that it seems of no practical use to you doesn't make it useless for everyone else. I think this issue should stay opened, because somebody (maybe even me) may create pull request for this.

It's a valid feature, it's implemented in tmux (this meant to show that people have been bothered enough not only to ask for this feature, but also enough for it to be implemented, there is also analogous feature request for emacs (not from me). Some people have bad eyesight and use big font, so this makes screen real estate more valuable. Also there are different activities when editing text (editing in insert mode, viewing, editing with different commands). I think on average people just move around about half the time. When moving around there is no point in showing empty command line.

So while this is a minor issue, it exists and have a right to continue to exist.

**Comment by chrisbra** [comment]
You are comparing apples and oranges here. For things like tmux, you don't really need a commandline, while for Vim you most likely need to use ex-cmds sooner or later (for e.g. help or quitting). I am for example almost never using the tmux commandline (and also did not for screen)

**Comment by nuko8** [comment]
Actually, the usefulness of the command line partly comes from the fact that it usually shows nothing there; that way once it shows something there, people will easily be aware that something remarkable is going on there at the moment in an un-distracting manner.  Otherwise, if the appearance of a text editor were to change suddenly in accordance with some necessity, people would have to examine the given visual change with care and make sure for themselves whether or not the change was due to any actual changes in buffers so that they would make an appropriate decision on what to do next.

**Comment by purpleP** [comment]
@chrisbra Of course you are right, in tmux command line is less used, but your argument about `using it eventually` is wrong, because there is no UX problem with showing command line on demand for such use case.

@nuko8 However is right, and I thought about it already, ideally the messages showed in command line should have some way of attracting attention. Like changing background or something like that. But I personally would use it even without it.

**Comment by vim-ml** [comment]
I suppose that allowing setting 'cmdheight' to zero (without changing
the default of 1, and if zero, hiding the commandline whenever it's
empty) would be a minimal and backward-compatible change to Vim.

However, not only would it add jumpiness whenever Vim has a message to
display, it would be inoperant in any mode other than Normal, since
the command-line space is used to display the mode at far left.

It would also mean that the (IMHO annoying) hit-enter-prompt (q.v.)
would be triggered for every single message that Vim would want to
display. Me, I use instead

        if &cmdheight < 2
          set ch=2
        endif

in my vimrc in order to avoid the most frequent cases of
hit-enter-prompt display. So I'm not much in favour of such a change
but I wouldn't vote against its being implemented.

In order for such an option to be "maximally" effective, you would
probably also want to set noshowcmd and possibly also laststatus=0 and
noruler: see

        :help 'showcmd'
        :help 'laststatus'
        :help 'ruler'

since all of these options affect either the command-line or the real
estate available for editing.


Best regards,
Tony.

**Comment by purpleP** [comment]
@vim-ml I don't understand the point about hit enter prompt in messages. Can you expand on that?

**Comment by ZyX-I** [comment]
@purpleP @vim-ml is a bot which reposts messages from mailing list. AFAIR it was supposed to post author name somewhere as well, but as you see it does not. Specifically Tony Mechelynck has his first name at the bottom.

For hit-enter prompts: try doing `:echo "a"`, `:echo "a\nb"`, `:echo "a\nb\nc"`. You will see that hit-enter prompt appears each time number of lines echoed is greater then &cmdheight.

**Comment by vimpostor** [comment]
FYI neovim is currently implementing this with `cmdheight=0`: https://github.com/neovim/neovim/pull/16251

I think it works beautifully and it is awesome to have one extra line of buffer content available due to the cmdline being hidden.
Neovim is handling the "messages hit-enter" problem by introducing a `msgfunc` option, where the user can define a new function that displays all messages (like `:echo test`), effectively moving the problem handling to the user.

I think it is time to reconsider this in vim. When the user is typing something into cmdline, the cmdline can simply be overlayed over the last line of buffer content and the messages problem can be taken care of with a `msgfunc` implementation (which can be completely independent of `cmdheight=0`)

**Comment by Shougo** [comment]
> FYI neovim is currently implementing this with cmdheight=0: https://github.com/neovim/neovim/pull/16251

It is merged now.  But the porting seems hard.
Because neovim has `cmdheight=0` base implementation for external GUI.  Vim has not.

**Comment by Shougo** [comment]
I have ported `set cmdheight=0` to Vim.

**Comment by vimpostor** [comment]
> I have ported `set cmdheight=0` to Vim.

You are an absolute legend! :tada:
